lib/Core/BuildEngine.cpp

```diff
@@ -1329,7 +1329,7 @@ class BuildEngineImpl : public BuildDBDelegate {
     // The RHS of the mapping is actually ignored, we use the StringMap's ptr
     // identity because it allows us to efficiently map back to the key string
     // in `getRuleInfoForKey`.
-    auto it = keyTable.insert(std::make_pair(key, KeyID::novalue())).first;
```

Comment:
Should the key table just map based on the key?

Reply:
keyTable is a StringMap, and as such, wants a StringRef for the key. Seems a reasonable choice of data structure for what we are doing with it right now.   We could split usages that want a StringRef out from a std::string&.  That could allow us to swap out the internals and differentiate performance for those that actually need a std::string from those that just need the ref.  That said, though, it can also be done later, if/when we want to do that.

Reply:
Makes sense